--- a/krping/krping.c	2018-12-11 05:51:43.000000000 -0800
+++ b/krping/krping.c	2019-02-07 17:47:00.611939897 -0800
@@ -46,7 +46,7 @@
 #include <linux/pci.h>
 #include <linux/time.h>
 #include <linux/random.h>
-#include <linux/sched/signal.h>
+#include <linux/signal.h>
 #include <linux/proc_fs.h>

 #include <asm/atomic.h>
@@ -657,7 +657,7 @@
 	int ret;
 	struct ib_cq_init_attr attr = {0};

-	cb->pd = ib_alloc_pd(cm_id->device, 0);
+	cb->pd = ib_alloc_pd(cm_id->device);
 	if (IS_ERR(cb->pd)) {
 		printk(KERN_ERR PFX "ib_alloc_pd failed\n");
 		return PTR_ERR(cb->pd);
